pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,153 +5,153 @@ package watchers
-							if oldCN.DeepEqual(ciliumNode) &&
-								comparator.MapStringEquals(oldCN.ObjectMeta.Labels, ciliumNode.ObjectMeta.Labels) &&
-								comparator.MapStringEquals(oldCN.ObjectMeta.Annotations, ciliumNode.ObjectMeta.Annotations) {
-								equal = true
-								return
-							}
-							if k8s.IsLocalCiliumNode(ciliumNode) {
-								return
-							}
-							n := nodeTypes.ParseCiliumNode(ciliumNode)
-							k.nodeDiscoverManager.NodeUpdated(n)
-							k.K8sEventProcessed(metricCiliumNode, resources.MetricUpdate, true)
-						}
-					}
-				},
-				DeleteFunc: func(obj interface{}) {
-					var equal bool
-					defer func() { k.K8sEventReceived(apiGroup, metricCiliumNode, resources.MetricDelete, true, equal) }()
-					ciliumNode := k8s.CastInformerEvent[cilium_v2.CiliumNode](obj)
-					if ciliumNode == nil {
-						return
-					}
-					n := nodeTypes.ParseCiliumNode(ciliumNode)
-					k.nodeDiscoverManager.NodeDeleted(n)
-				},
-			},
+		var synced atomic.Bool
+		stop := make(chan struct{})
+
+		k.blockWaitGroupToSyncResources(
+			stop,
 			nil,
+			func() bool { return synced.Load() },
+			apiGroup,
 		)
-		isConnected := make(chan struct{})
-		// once isConnected is closed, it will stop waiting on caches to be
-		// synchronized.
-		k.blockWaitGroupToSyncResources(isConnected, swgNodes, ciliumNodeInformer.HasSynced, apiGroup)
-
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = ciliumNodeStore
-		k.ciliumNodeStoreMU.Unlock()
-
-		once.Do(func() {
-			// Signalize that we have put node controller in the wait group
-			// to sync resources.
-			asyncControllers.Done()
-		})
 		k.k8sAPIGroups.AddAPI(apiGroup)
-		go ciliumNodeInformer.Run(isConnected)
 
-		<-kvstore.Connected()
-		log.Info("Connected to key-value store, stopping CiliumNode watcher")
+		// Signalize that we have put node controller in the wait group to sync resources.
+		once.Do(asyncControllers.Done)
 
-		// Set the ciliumNodeStore as nil so that any attempts of getting the
-		// CiliumNode are performed with a request sent to kube-apiserver
-		// directly instead of relying on an outdated version of the CiliumNode
-		// in this cache.
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = nil
-		k.ciliumNodeStoreMU.Unlock()
+		// Set forceCiliumNodeFromApiServer to false so that any attempts of getting the
+		// CiliumNode go through the Resource[T] cache.
+		k.forceCiliumNodeFromApiServer.Store(false)
 
-		close(isConnected)
+		// derive another context to signal Events() in case of kvstore connection
+		eventsCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
 
-		k.cancelWaitGroupToSyncResources(apiGroup)
-		k.k8sAPIGroups.RemoveAPI(apiGroup)
-		// Create a new node controller when we are disconnected with the
-		// kvstore
-		<-kvstore.Client().Disconnected()
+		go func() {
+			defer close(stop)
 
-		log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+			events := k.resources.CiliumNode.Events(eventsCtx)
```

Comment:
AFAICT, there's a notable difference compared to the previous approach. Earlier, we stopped the informer whenever the connection to the kvstore was established, and restarted it only in the event of a disconnection. With this implementation, instead, we would never stop it, regardless of whether the context passed to Events gets canceled earlier (that only breaks the loop, with the informer continuing to run in background until the stop hook of the corresponding resource is executed.

This would defeat the point of watching node updates from the kvstore, and I think it would be worth additional discussion whether this is something that we really want to get rid of at this point, due to the possible scalability implications (especially because CiliumNode CRs can be pretty chatty in case of ENI IPAM mode). Not that the previous implementation was really great from this point of view, as in case of kvstore disconnection it would cause a full relist form k8s, to likely just discard everything immediately afterwards on kvstore reconnection (my understanding is that it was required to break the :chicken::egg: dependency if the kvstore runs in pod network).

The other point is that, if we preserve the synchronization from the kvstore, then the EGW should not watch directly the CiliumNode CRs, as that prevents it from working in the kvstore/clustermesh case (at least as a contributing factor, not necessarily the only one).

/cc @marseel

Reply:
I think this PR is still an improvement compared to the previous implementation. From the agent point of view, two different informers were active at the same time for the same resource. One of the informer, namely the one started from Resource[T] used in the Egress Gateway manager, would never have been stopped in the previous version too.

With the proposed changes, it is true that cancelling the context passed to `Events` won't stop the Resource[T] informer either, but that is expected, since it is still needed to pump the updates to the Egress Gateway manager.

If we decide to avoid depending on the Resource[CiliumNode] (or unsubscribe from it at runtime) in the EG manager, then what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.

Reply:
> From the agent point of view, two different informers were active at the same time for the same resource. One of the informer, namely the one started from Resource[T] used in the Egress Gateway manager, would never have been stopped in the previous version too.

Hmm, that's true, but only if the Egress Gateway is enabled, and kvstore is not enabled (the two are not compatible anyway). When kvstore is enabled, instead, this PR is introducing a "regression"  there, because the expectation that the informer gets stopped after connecting to the kvstore does no longer hold true.

I personally don't have a strong opinion on what approach is best, each has its own advantages and disadvantages. Although my understanding is that the Egress Gateway should not watch the CiliumNodes in the first place, but use an higher-level abstraction to be compatible both with CRD and kvstore mode.

Reply:
And the same behavioral change would be also caused for CiliumEndpoints by [this other commit](https://github.com/cilium/cilium/pull/29249/commits/6386f9490d168cdc9572f37721cbbb731788c497).

Reply:
The fact that this issue may affect more than one resource types makes me think that it should be addressed in the underlying package. Let's see if adding this to Resource[T]:

> what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.

is feasible. I'm gonna try that in another PR :+1:

Reply:
> what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.

I'm not really sure whether this is really feasible when used in combination with the Store API, as after stopping it would contain stale data only. And tat could break other consumers in many unexpected ways.

More in general, maybe this thread could be a good topic of discussion during today's community meeting. WDYT?

Reply:
> > what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.
> 
> I'm not really sure whether this is really feasible when used in combination with the Store API, as after stopping it would contain stale data only. And that could break other (unaware) consumers in many unexpected ways.

Actually the Store API marks the informer as needed and block the caller till the cache is synced again. I was thinking about keeping this behavior while introducing the change above.

> More in general, maybe this thread could be a good topic for discussion during today's community meeting. WDYT?

That would be great, super happy to discuss this in more detail. :+1:

Reply:
> Actually the Store API marks the informer as needed and block the caller till the cache is synced again. I was thinking about keeping this behavior while introducing the change above.

Yep, my point is that stopping the informer when events are no longer observed, but keeping it if the anyone gathered the store, would defeat the whole idea of stopping the informer (as for instance we always retrieve the node store IIRC). And it would also introduce more uncertainty on whether a given resource is actually being watched or not (as the stopping conditions would also play a role, in addition to the starting ones).

> That would be great, super happy to discuss this in more detail. 👍

:+1:  I'll add the topic to the agenda

Reply:
> the Egress Gateway is enabled, and kvstore is not enabled (the two are not compatible anyway)

Most likely I was wrong about this incompatibility, because the CiliumNode CRs are present (and updated) regardless of the kvstore mode. While the point still holds in case of Cluster Mesh.

Reply:
tldr at the end

> I'm not really sure whether this is really feasible when used in combination with the Store API, as after stopping it would contain stale data only.

I think it would make sense to "force" consumers to comply with this requirement - consumers should depend on events to trigger updates and not rely on the store only  - so if they are consumers, they should have a channel for receiving events (and if they close it they should know that store will be obsolete) etc and respect this assumption (I understand this might not be an easy change though).

I didn't have a strong preference before if we should/shouldn't stop informers when switching from CRD to kvstore in this case. The reasoning is that:
- if we close informer, then once kvstore somehow gets broken, all of the agents will spam the k8s control plane simultaneously, which is not great and will result in degraded performance for a long time (I am guessing possibly tens of minutes) - still not as bad contributor to overall load as Endpoints
- if we don't close the informer, we still put some pressure on k8s control plane, probably insignificant in most of the cases (unless we have ENI mode as mentioned before for the Node CRD case)
- also, one side effect of having watches opened and not closing informer will be that update requests to Node CRDs will get slower (k8s takes into account the number of watches opened to specific resource in the context of throttling update requests, reducing throughput of updates up to 10x). 

This makes me think that we should have a general guideline on when to close/not close informers in the case of kvstore as it's super non-obvious. 
I would suggest:
- Frequent Node updates in ENI mode is a separate issue that should be addressed regardless of this PR. I think the best solution would be to fix ENI issue + keep the watch open. But until we fix the ENI issue, closing the informer would make it at least as good as before. 
- We should do some retries after disconnecting from kvstore before trying to get data from CRD to protect the k8s control plane (I will open a separate issue for that) - this is especially important for other resources like Endpoints/Pods (we watch all Pods when Endpoint CRD is disabled, that's a separate issue) which have much higher cardinality and churn.
- It should be fine to keep watches for all other resources - local Endpoints, Identities, Nodes, ... in general etc So basically only avoiding Endpoints/Pods.

So now the question becomes, what's the purpose of having kvstore:
- Offload Endpoint load/churn to kvstore in k8s clusters (still we have the same issue with switching between CRD and kvstore for this resource, I think we would need to figure out a better solution for it)
- Cilium deployments without a k8s control plane
- Expose data in case of some of the clustermesh deployments - which hopefully we would like to get rid of in long-term
- kvstore identity allocation mode - I don't think it brings many benefits in k8s clusters(and doesn't allow to use "regular" clustermesh control plane, which forces users to stick with the previous point), but is required in cilium deployments without a k8s control plane
- allows to disable CiliumEndpoint CRD in k8s clusters - I believe this actually brings more harm than benefits currently as it forces agents to list/watch all pods until connecting to kvstore (similar issue to the first point). 
- ... something else I missed?

tldr
- For this PR, closing informer once connected to kvstore is fine and as good as it was before. It would be also fine to not close the informer if we fix the ENI issue. 
- We should improve retries of disconnects from kvstore to reduce the load on the k8s control plane in case of Etcd failures
- long-term offloading the k8s control plane with Nodes/Identities churn (after fixing the ENI issue) shouldn't be a priority for kvstore deployments and we should probably focus on better handling of switches between CRD/kvstore in case of Endpoints (and Pods when Endpoints CRD is disabled).
- we **really** need some good abstraction for fetching and storing data from/to CRDs/kvstore/clustermesh (stateDB?)